### DISCUSSION TITLE
Allow specifying IP address for Express server to listen on

### DISCUSSION THREAD

**Discussion by codycraven** [remix-run/remix#2320] [Proposals]
Developers have a variety of environments that may make routing challenging (running code in VM, containers, WSL2, etc). A fairly consistent solution to these environment specific challenges is to attach listeners to IP address 0.0.0.0 instead of a specific IP (which may only be accessible in a specific scope).

To lower the barrier of entry for these developers, an environment variable could be added to specify the IP address to listen (just like the PORT can be specified).

Example:
- Update: https://github.com/remix-run/remix/blob/c4dd5f128ca65f97c114621b2c8140bccaaaacda/packages/remix-dev/cli/commands.ts#L185-L195
- To: 
    ```ts
    let host = process.env.HOST ?? (Object.values(os.networkInterfaces()) 
      .flat() 
      .find((ip) => ip?.family == "IPv4" && !ip.internal)?.address); 
 
     if (!host) { 
      host = "localhost"; 
    } 
 
    server = app.listen(port, host, () => { 
      console.log(`Remix App Server started at http://${host}:${port}`); 
    }); 
    ```
    Then the developer could start the dev server as: `HOST=0.0.0.0 npm run dev`

Note, if it's a good idea to implement this, it should probably be done for remix-serve/cli.ts as well.

I'm happy to open a PR if this is wanted.

**Comment by kentcdodds** [comment]
Thanks for taking the time to open this. Will it help at all that we're publishing a change today to show both the `localhost` URL as well as the local IP address?

**Reply by codycraven** [reply]
This won't help for any situations where the process's network is directly inaccessible from the host — such as with containers.

**Reply by kentcdodds** [reply]
Gotcha. I think I'm up for this personally. I don't understand how using `0.0.0.0` will help in those situations though 🤔

**Reply by codycraven** [reply]
The container runtime can expose ports from the virtual network to the host's network, however the requests will come to an IP address that is not the container's primary IP address. Listening on 0.0.0.0 allows the server (express/node in this case) to handle a request to the bound port regardless of the network interface it is received from.

**Reply by kentcdodds** [reply]
I was unaware of this! Is there a reason we shouldn't just default to that?

**Reply by codycraven** [reply]
0.0.0.0 is definitely less secure (you could expose your dev site to everyone in a coffee shop for example), so really should only be used in these virtual network scenarios (or on trusted private networks).

**Reply by kentcdodds** [reply]
Got it! Thank you for explaining that. I think a pull request for this would be welcome 👍 Thanks!
